### PR TITLE
fix(core): Fix hardcoded `aud` claim in MCP OAuth tokens

### LINKED ISSUE #30500: Bug: Instance-level MCP server issues OAuth tokens with hardcoded aud:"mcp-server-api", ignoring the `resource` parameter (RFC 8707 violation)
### Bug Description

n8n's Instance-level MCP (`/mcp-server/http`) issues OAuth access tokens whose JWT `aud` claim is hardcoded to the string `"mcp-server-api"`, regardless of the `resource` parameter the client sends in the authorization request (per RFC 8707 / [MCP Authorization spec](https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization)).

Spec-compliant MCP clients (e.g. Claude.ai's custom-connector flow) validate that `aud` matches the canonical resource URL of the MCP server. When that validation fails they silently discard the token. The complete OAuth dance succeeds through `POST /mcp-oauth/token → 200`, but the client then makes zero further calls and reports an authorization failure.

This effectively blocks any spec-compliant MCP client from connecting to n8n's Instance-level MCP via the OAuth flow.

### To Reproduce

1. Enable Instance-level MCP in n8n.
2. Run the standard MCP OAuth flow against the server, sending `resource=https://<your-n8n>/mcp-server/http` in the authorize request. (Easy reproduction: `curl` walkthrough below.)
3. Decode the issued JWT — `aud` is `mcp-server-api`, not the resource URL.

Reproducer using a confidential client previously registered via `/mcp-oauth/register`:

```bash
# 1. Register a client (one-time)
curl -X POST https://<n8n>/mcp-oauth/register \
  -H 'Content-Type: application/json' \
  -d '{
    "client_name": "diagnostic",
    "redirect_uris": ["https://example.invalid/cb"],
    "grant_types": ["authorization_code","refresh_token"],
    "response_types": ["code"],
    "token_endpoint_auth_method": "client_secret_basic"
  }'
# Returns client_id + client_secret + refresh_token mechanism

# 2. After completing an authorize → token flow, decode the resulting access_token.
# OR use refresh_token grant with resource param explicitly set:
curl -X POST https://<n8n>/mcp-oauth/token \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -d "grant_type=refresh_token&refresh_token=<RT>&client_id=<CID>&client_secret=<SECRET>&resource=https://<n8n>/mcp-server/http"

# Decode payload (middle segment) of returned access_token:
# {
#   "sub": "...",
#   "aud": "mcp-server-api",          ← hardcoded; should be the resource URL
#   "client_id": "...",
#   "jti": "...",
#   "iat": ...,
#   "exp": ...,
#   "meta": {"isOAuth": true}
# }
```

### Expected behavior

Per [RFC 8707 §2.2](https://datatracker.ietf.org/doc/html/rfc8707#section-2.2) and the MCP Authorization spec, the issued access token's `aud` claim should equal the `resource` URI provided in the request (or, when multiple resources are supported, contain the requested resource). Something like:

```json
"aud": "https://n8n.example.com/mcp-server/http"
```

This allows the client to verify the token was issued for its intended resource, preventing token-confused-deputy attacks.

## Actual behavior

`aud` is the literal string `"mcp-server-api"`, regardless of the request. The `resource` parameter sent by the client is parsed (the authorize endpoint accepts it without error) but does not flow into the issued token.

### Debug Info

# Debug info

## core

- n8nVersion: 2.17.5
- platform: npm
- nodeJsVersion: 24.14.0
- nodeEnv: undefined
- database: sqlite
- executionMode: regular
- concurrency: -1
- license: enterprise (production)
- consumerId: 68499a19-5350-4193-a3a9-9bd579e3d2ed

## storage

- success: all
- error: all
- progress: false
- manual: true
- binaryMode: filesystem

## pruning

- enabled: true
- maxAge: 336 hours
- maxCount: 10000 executions

## client

- userAgent: mozilla/5.0 (windows nt 10.0; win64; x64) applewebkit/537.36 (khtml, like gecko) chrome/148.0.0.0 safari/537.36 edg/148.0.0.0
- isTouchDevice: false

## security

- secureCookie: false

Generated at: 2026-05-15T06:57:52.203Z

### Operating System

Debian GNU/Linux 13 (trixie)

### n8n Version

2.17.5

### Node.js Version

24.14.0

### Database

SQLite (default)

### Execution mode

main (default)

### Hosting

self hosted

### PR DESCRIPTION
**Pull request by sinehypernova-0718**
## Summary

The Instance-level MCP OAuth server always signed access tokens with a hardcoded `"aud": "mcp-server-api"` claim. However, spec-compliant MCP clients expect the token audience to match the resource URL they originally requested. Since the two values never matched, clients would silently reject the token, causing the MCP connection to fail even after a successful OAuth flow.

With this fix, MCP OAuth access tokens no longer use a hardcoded `aud` claim. Instead, the audience is now derived from the `resource` parameter whenever the client provides one. Tokens using the old audience are still accepted and continue to verify correctly, so existing integrations remain unaffected. This also fixes the mismatch between the advertised protected-resource URL and the issued token audience (#30054).

This change also adds support for resource indicators throughout the MCP OAuth flow. A new resource column has been added to authorization codes, the resource value is now propagated through the session and consent layers, and resource indicators are validated during both OAuth authorization and token exchange. The issued JWT audience is then derived from these resources, enabling proper audience-aware token verification.

### The fix

* **Dynamic `aud` claim** – The `resource` parameter from the authorization request is now carried through the full OAuth flow (session → consent → authorization code → token generation). If a valid resource is provided, it is used as the token’s `aud` claim. If no resource is provided, the system falls back to the instance’s canonical MCP resource URL (since `generateTokenPair` now uses `resource ?? getResourceUrl()`).

* **Backward-compatible token verification** – The token verifier now accepts both the canonical resource URL and the legacy audience value. This ensures that older tokens continue to work without requiring any migration or breaking existing integrations.

* **Resource validation** – Requests that include a `resource` value which does not match the server’s configured MCP endpoint now receive a **400 `invalid_target` error** (as defined in RFC 8707). This prevents clients from obtaining tokens for unrelated or third-party audiences.

## How it works (high‑level flow)

1. **Authorize endpoint** – The `resource` parameter is now extracted directly from the authorization request (using `Reflect` for SDK compatibility). It is then validated against `<instanceBaseUrl>/mcp-server/http` and stored inside the OAuth session JWT for later stages of the flow.

2. **Consent screen** – During consent handling, the resource value is read back from the session and persisted in the `oauth_authorization_codes` table through a new nullable `resource` column.

3. **Token exchange** – When exchanging the authorization code for tokens, the stored resource is forwarded to `McpOAuthTokenService.generateTokenPair`, which sets the JWT audience as `aud = resource ?? getResourceUrl()` (the canonical instance URL)..

4. **Token verification** – The verification middleware now builds the expected audience dynamically using `UrlService.getInstanceBaseUrl()` and passes it into `verifyAccessToken`. The verifier accepts either the canonical MCP resource URL or the legacy `"mcp-server-api"` audience through an allowed audience list: `[canonicalResource, 'mcp-server-api']`.

5. **Refresh token rotation** – Resource information is also preserved during refresh token rotation. If the client does not provide a resource during refresh, the flow falls back to the legacy audience behavior for compatibility.

## Backward compatibility

| Scenario | `aud` in token | Accepted? |
|----------|----------------|-----------|
| No `resource` parameter | `"https://<instance>/mcp-server/http"` (canonical) | ✅ Yes |
| Valid `resource` parameter | `"https://<instance>/mcp-server/http"` | ✅ Yes |
| Legacy token (pre‑migration) presented to middleware | `"mcp-server-api"` | ✅ Yes |
| Refresh without `resource` | Uses Canonical resource URL| ✅ Yes |
| Refresh with `resource` | Uses resource URL | ✅ Yes |

**No existing workflows or integrations are affected. The legacy audience value continues to be supported for both token issuance and token verification whenever no resource parameter is provided.**

> [!NOTE]
> After PR #30055 merged, the default audience for new tokens became the canonical resource URL. The legacy `"mcp-server-api"` string is now only used for backward‑compatible token verification.

## Database migration

A new migration (`1784000000024-AddResourceToOAuthAuthorizationCodes.ts`) adds a nullable `resource` column to the `oauth_authorization_codes` table. The migration is reversible (implements `ReversibleMigration` using the `schemaBuilder` API), tested on SQLite and PostgreSQL 16, and registered for both database engines.

### Core OAuth logic
- `packages/cli/src/modules/mcp/mcp-oauth-service.ts` – Resource extraction, validation, and error handling.
- `packages/cli/src/modules/mcp/mcp-oauth-token.service.ts` – Dynamic `aud`, multi‑audience verification, safe claim extraction.
- `packages/cli/src/modules/mcp/mcp-oauth-consent.service.ts` – Passes resource from session to authorization code.
- `packages/cli/src/modules/mcp/mcp-oauth-authorization-code.service.ts` – Persists resource; atomic code consumption.
- `packages/cli/src/modules/mcp/mcp-server-middleware.service.ts` – Verifies against instance resource URL.
- `packages/cli/src/modules/mcp/oauth-session.service.ts` – Extended session payload.

### Data layer
- `packages/cli/src/modules/mcp/database/entities/oauth-authorization-code.entity.ts` – New `resource` column.
- `packages/@n8n/db/src/migrations/common/1784000000024-AddResourceToOAuthAuthorizationCodes.ts` – Migration.
- `packages/@n8n/db/src/migrations/postgresdb/index.ts`, `packages/@n8n/db/src/migrations/sqlite/index.ts` – Migration registrations.

**Manual verification steps:**
1. Register an MCP OAuth client.
2. Authorize with `resource=https://<n8n-instance>/mcp-server/http`.
3. Exchange code for token → decode JWT → `aud` matches the resource URL.
4. Use the token against an MCP endpoint → accepted.
5. Repeat without `resource` → token uses the instance’s canonical resource URL as `aud` → still accepted.
6. Try an invalid `resource` → `400` error.

> [!IMPORTANT]
> New middleware must be deployed before or at the same time as the new token service. During a rolling restart, a token issued by an updated pod will carry the canonical resource URL as `aud`; an older pod that only accepts `'mcp-server-api'` will reject it. To avoid transient 401s, perform a hard cutover (all instances updated simultaneously) or ensure the new verification code is live before the new token‑issuing code.
## Related Linear tickets, Github issues, and Community forum posts

- Fixes #30500
- Linear Ticket ID: [GHC-8363](https://linear.app/n8n/issue/GHC-8363)
- would have fixed, currently related: #30054 / PR #30055
- Linear Ticket ID: [GHC-8221](https://linear.app/n8n/issue/GHC-8221)
- related: #30733 (MCP Client Tool OAuth2 flow omits RFC 8707 `resource` parameter, the client‑side counterpart of this server‑side fix)/ PR #30924
- Linear Ticket ID: [GHC-8431](https://linear.app/n8n/issue/GHC-8431)
## Review / Merge checklist

- [x] I have seen this code, I have run this code, and I take responsibility for this code.
- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md))
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included.
- [ ] PR Labeled with `Backport to Beta`, `Backport to Stable`, or `Backport to v1` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

**Comment by sinehypernova-0718** [comment]
@codex are your suggestion comments addressed in the latest commit 85c6d8478e744b2865e0a26677f350cfa797f216 if yes, then resolve the conversation of the outdated comments and if you have new suggestions comment them here after review

**Comment by sinehypernova-0718** [comment]
@codex review the latest change stack

**Comment by sinehypernova-0718** [comment]
## Backward Compatibility Notes 


**Adding detail beyond the PR description for review / future reference.**
### Token Issuance Scenarios

| Scenario | `aud` in token | Accepted? | Notes |
|----------|----------------|-----------|-------|
| No `resource` parameter (authorize) | `"https://<instance>/mcp-server/http"` (canonical) | ✅ Yes | Defaults to the instance’s canonical resource URL |
| Valid `resource` parameter (authorize) | `"https://<instance>/mcp-server/http"` | ✅ Yes | RFC 8707 compliant |
| Refresh **without** `resource` (regardless of original auth) | `"https://<instance>/mcp-server/http"` (canonical) | ✅ Yes | Rotated token always uses the canonical URL when no resource is given |
| Refresh **with** `resource` | `"https://<instance>/mcp-server/http"` | ✅ Yes | Consistent with the authorize flow |
| Authorization code with trailing slash `.../mcp-server/http/` | `"https://<instance>/mcp-server/http"` (normalized) | ✅ Yes | Trailing slash stripped before storage and token generation |
| Token endpoint provides `resource` conflicting with stored code | N/A (token rejected) | ❌ No | `invalid_target` returned; code NOT consumed (client can retry) |

### Token Verification Scenarios

| Scenario | Token `aud` | Middleware expected `aud` | Accepted? | Notes |
|----------|-------------|---------------------------|-----------|-------|
| Legacy token (pre‑migration) presented | `"mcp-server-api"` | `"https://<instance>/mcp-server/http"` | ✅ Yes | Allowed audience array includes legacy |
| New token (with resource) presented | `"https://<instance>/mcp-server/http"` | `"https://<instance>/mcp-server/http"` | ✅ Yes | Exact match on primary audience |
| Token from instance with *different* base URL | `"https://old-instance.example.com/mcp-server/http"` | `"https://new-instance.example.com/mcp-server/http"` | ❌ No | Known limitation,  see migration notes below |
| Token with tampered `aud` (e.g., third-party URL) | `"https://attacker.example.com/mcp-server/http"` | `"https://<instance>/mcp-server/http"` | ❌ No | Rejected by JWT verification (not in allowed audiences) |
| Token from SDK `verifyAccessToken` call (no expected audience) | `"mcp-server-api"` | N/A (uses `verifyAccessToken(token)` without audience) | ✅ Yes | Legacy single-audience verification; SDK path unaffected |

### Deployment & Migration Scenarios

| Scenario | Impact | Mitigation |
|----------|--------|------------|
| Rolling deploy (old middleware + new token service) | New tokens issued with `aud: "https://..."`, old middleware expects `"mcp-server-api"` only | ❌ New tokens rejected during deploy window | Deploy middleware change first, or accept brief 401s |
| Rolling deploy (new middleware + old token service) | Old tokens issued with `aud: "mcp-server-api"`, new middleware accepts both | ✅ Works | Safe deployment order: middleware → token service |
| Database migration runs on existing data | `resource` column added as nullable; existing codes have `NULL` resource | ✅ No impact | `NULL` → `authRecord.resource ?? undefined` → falls back to **canonical** resource URL (not legacy) |
| Migration rollback (down migration) | `resource` column dropped | ✅ Clean | Reversible migration; tokens with resource `aud` remain valid until expiry |
| Instance URL change (domain migration) | Canonical resource URL changes; existing tokens have old URL in `aud` | ❌ Existing tokens rejected | Document that re‑authorization is required after domain change |

> [!NOTE]
> These tables were updated after PR #30055 was merged into master. That PR replaced the hardcoded `"mcp-server-api"` audience with the canonical instance resource URL for all new token issuance. As a result, tokens are now minted with `aud = resource ?? getResourceUrl()` (the canonical URL), and the legacy `"mcp-server-api"` string is retained only for backward‑compatible token verification. The original version of this PR defaulted to the legacy string when no `resource` was provided; the final code reflects the combined behavior.

**Comment by sinehypernova-0718** [comment]
Hi! @RicardoE105  I saw you merged #30055, nice quick and effective fix! It definitely helps with the immediate issue, and I appreciate it.

I wanted to mention that I’ve been working on the same root issue (#30500) in PR #30558, and I think it might be worth a quick look since it goes a bit deeper into making instance-level MCP production-ready. #30055 replaces the hardcoded audience, but it still doesn’t use the client’s `resource` parameter, so it’s not fully RFC 8707 compliant yet.

That gap is already causing real issues too, #30733 shows the MCP Client Tool OAuth2 flow breaking against spec-compliant servers because the `resource` parameter is missing. The OAuth flow itself succeeds, but all API calls after that fail with `401 invalid_token`. So this isn’t just a spec detail anymore; it’s actively affecting users.

My PR #30558 tries to handle the full flow properly by:
- Capturing and validating the client’s `resource` parameter
- Passing it through the entire OAuth flow (authorize → consent → auth code → token)
- Using proper OAuth error codes (`invalid_target`, `invalid_grant`)
- Adding a reversible database migration to persist the resource
- Keeping full backward compatibility, so existing tokens continue working (see "Backward Compatibility Notes" in the comments)

Since instance-level MCP is still pretty new, I feel getting this right early could save a lot of pain later on. Not forcing or anything like that just a quick heads up from my side! If you get a chance to take a look at #30558, 

I’d really appreciate your thoughts on this!

Thank You for your considering!

**Comment by sinehypernova-0718** [comment]
Hey @RicardoE105, just a gentle ping in case this got buried.

Thanks!

**Comment by RicardoE105** [comment]
@sinehypernova-0718 thanks for the ping. Will try to have a look over the weekend.

**Comment by sinehypernova-0718** [comment]
Thanks! @RicardoE105, really appreciate it! Let me know if you need anything from my side!

**Comment by sinehypernova-0718** [comment]
Hey! Quick update, I pushed a few more commits for extra test coverage and edge-case handling, as you can see above.

I’ll update the Database Migration (timestamp, type, `.text.comment(...))` as mentioned in the migration guidelines/skill after review, since master has moved ahead and I’d rather regenerate it once everything else is settled.

Apologies for the disturbance!

**Comment by sinehypernova-0718** [comment]
Hey @RicardoE105!  Just checking in, a gentle ping once again. I know you must be busy Whenever you have a moment to take a look. 
Thanks again for your time! 🙂

File: packages/cli/src/modules/mcp/mcp-oauth-service.ts [hunk 1]
```diff
@@ -188,21 +210,61 @@ export class McpOAuthService implements OAuthServerProvider {
...
 			authorizationCode,
 			client.client_id,
 			redirectUri,
 		);
 
+		if (!authRecord) {
+			throw new OAuthError('invalid_grant', 'Invalid authorization code');
+		}
+
+		// Convert URL to string if provided
```
Comment (RicardoE105) [inline packages/cli/src/modules/mcp/mcp-oauth-service.ts]: // Convert URL to string if provided is a what-comment, the code already says that. Can drop.

File: packages/cli/src/modules/mcp/mcp-oauth-service.ts [hunk 2]
```diff
@@ -188,21 +210,61 @@ export class McpOAuthService implements OAuthServerProvider {
...
+		const resourceStr = resource?.toString();
+		const tokenResource = this.resolveAndValidateResourceIndicator(resourceStr);
+
+		// Resource reconciliation per RFC 8707:
+		//
+		// We reconcile the requested token resource with the resource bound to the authorization code:
+		// - If both are specified: they must match exactly, otherwise we reject the exchange.
+		// - If only the token request specifies a resource: we use it (client override).
+		// - If only the authorization code contains a resource: we fall back to it.
+		// - If neither is specified: we default to the legacy 'mcp-server-api' audience.
```
Comment (RicardoE105) [inline packages/cli/src/modules/mcp/mcp-oauth-service.ts]: This comment says "default to the legacy 'mcp-server-api' audience" but that's not what the code does. When neither resource is specified, finalResource is undefined, and generateTokenPair falls back to getResourceUrl() (the canonical URL), not LEGACY_MCP_AUDIENCE. The PR description's compat table reads the same way. Could you fix both so they match what the code actually does?

File: packages/cli/src/modules/mcp/mcp-oauth-service.ts [hunk 3]
```diff
@@ -225,18 +282,115 @@ export class McpOAuthService implements OAuthServerProvider {
...
+	 * `scopesOrResource` may be:
+	 * - `string[]`: scopes (intentionally treated as no resource, resulting in `undefined`)
+	 * - `string` or `URL`: resource indicator
+	 *
+	 * This follows OAuth 2.1 refresh-token behavior, where scopes are not used for token refresh.
+	 * The selected resource is normalized and validated via `resolveAndValidateResourceIndicator()`.
+	 *
+	 * @see exchangeAuthorizationCode For initial authorization code exchange.
+	 */
 	async exchangeRefreshToken(
```
Comment (RicardoE105) [inline packages/cli/src/modules/mcp/mcp-oauth-service.ts:269]: This diverges from the SDK contract, which is (client, refreshToken, scopes?: string[], resource?: URL). The type guards happen to do the right thing but it's confusing and easy to break later. Can you match the SDK shape? (client, refreshToken, _scopes, resource) is enough.

File: packages/cli/src/modules/mcp/mcp-oauth-service.ts [hunk 4]
```diff
@@ -225,18 +282,115 @@ export class McpOAuthService implements OAuthServerProvider {
...
+	 * Returns undefined if the 'resource' parameter is missing, null, or undefined.
+	 * Throws InvalidResourceIndicatorError if a resource is present but is not a string or URL.
+	 *
+	 * @see https://datatracker.ietf.org/doc/html/rfc8707
+	 * @see https://spec.modelcontextprotocol.io/specification/2025-03-26/basic/authorization/
+	 */
+	private getResourceIndicatorFromAuthorizationParams(
+		params: AuthorizationParams,
+	): string | undefined {
+		const hasResource = Reflect.has(params, 'resource');
```
Comment (RicardoE105) [inline packages/cli/src/modules/mcp/mcp-oauth-service.ts]: The Reflect indirection is overkill, the SDK's AuthorizationParams in the version we're on already declares resource?: URL (see provider.d.ts). A plain params.resource access with a string | URL | undefined check works and is much easier to read.

File: packages/cli/src/modules/mcp/mcp-oauth-token.service.ts [hunk 5]
```diff
@@ -248,4 +266,57 @@ export class McpOAuthTokenService {
...
+	 * BACKWARD COMPATIBILITY: We accept both the canonical resource URL and the legacy
+	 * 'mcp-server-api' audience during this transition period.
+	 *
+	 * TODO: Remove the fallback loop once all legacy tokens in the wild expire
+	 * (based on refresh token lifespan). Legacy tokens minted prior to n8n v2.19
+	 * had a hardcoded 'mcp-server-api' audience.
+	 */
+	private verifyJwtWithAllowedAudiences(token: string, audiences: string[]): unknown {
+		if (audiences.length === 0) {
+			throw new Error('At least one audience is required');
```
Comment (RicardoE105) [inline packages/cli/src/modules/mcp/mcp-oauth-token.service.ts]: Unreachable, the only caller hardcodes a non-empty list. Can drop.

**Comment by sinehypernova-0718** [comment]
Thank you so much for the thorough review, @RicardoE105, I'm currently away from my setup, but I'll go through each of your suggestions and address them in about an hour. Excited to get this over the finish line. Thanks again!

File: packages/cli/src/modules/mcp/mcp-oauth-service.ts (hunk 2 above)
Comment (sinehypernova-0718) [reply packages/cli/src/modules/mcp/mcp-oauth-service.ts]: I might be going blind here 😅 but I re-checked both the PR description notes and the Extra backward-compat notes (the ones posted as a comment), and they look correct to me, the table already says `https://<instance>/mcp-server/http` (canonical) for the “No resource parameter” row.
The inline code comment was wrong though, that’s my bad. Leftover from an older draft. I’ve fixed it in the latest commit. Will push it shortly!

If there’s another stale spot I missed, could you point me to the exact line? Want to make sure everything matches up. Thanks!

**Comment by sinehypernova-0718** [comment]
@RicardoE105 All review comments addressed, stale comment updated, `exchangeRefreshToken` now matches the SDK contract, Reflect replaced with direct access, unreachable guard removed. The DB migration has also been regenerated with a current timestamp and `.text.comment(...)`. Let me know if there's anything else you'd like adjusted.

Edit: Build, Lint, and Typecheck passes locally too. 😅

Thanks for the review!

**Comment by sinehypernova-0718** [comment]
Hey @RicardoE105, I know I've been tagging you a fair bit and I don't want to be pushy at all 😅 I understand you might be busy, just wanted to mention that all the review comments are addressed whenever you have a moment. No rush whatsoever, and thanks again for the thorough review!

Edit: Regarding the DB migration, it seems there are conflicts once again with master, I’ve had already ran the migration update earlier along with you review comments, but  this time I’ll wait for your go‑ahead before regenerating it again, so there aren’t any timestamp or ordering issues at merge. 

Thanks!

**Comment by RicardoE105** [comment]
@sinehypernova-0718 set some time to review it tomorrow

**Comment by sinehypernova-0718** [comment]
Hey @RicardoE105, it’s getting late at night here, saw the CI checks and 1 lint failure is from my change and the E2E tests seem to cancel out of nowhere. I’ll push all the required fixes tomorrow after your review. Thanks for your time!

Heading off now, catch you tomorrow!

**Review by RicardoE105 — DISMISSED** [review]

**Comment by RicardoE105** [comment]
@cubic-dev-ai review.

**Review by RicardoE105 — APPROVED** [review]
